Spring cleaning with multiarch support

amotl opened the pull request:
Hi there,

along the lines of #113, @bogde and @el3ctrician approved the idea of a spring cleaning of this library. 

So, we would like to share what we have now: Different pieces of code and suggestions from various people have been incorporated, I will try to add references to the relevant issues on behalf of the next comment to keep this introduction shorter.

As I don't actually have any hardware on my workbench, I am humbly asking the community (all of you ;]) about whether you would like to give the revamped library a testdrive on a real iron? I will be happy to receive any feedback about the outcome.

While we are discussing this topic at [1], the work-in-progress branch is located at [2] and might receive further updates as we go. We hope you like it.

With kind regards,
Andreas.

[1] https://community.hiveeyes.org/t/improving-the-canonical-hx711-library-for-arduino/539
[2] https://github.com/hiveeyes/HX711/tree/spring-cleaning

amotl commented:
This actually bundles many suggestions floating around here since quite some time. We hope to have assembled them carefully into something which will be more useful for the community again by making people struggle less when starting with this library on more contemporary and powerful MCUs like ESP, SAMD, STM32 and others.

## This branch directly incorporates 
parts from these pull requests, selected by cherry-picking.
- [x] AVR interrupt safety, see #62 
- [x] ESP32 support, see #75
- [x] Teensy 3.x support, see #96 

Also, it will probably contain different bits and pieces from issue requests which have not been pulled in using PRs. We tried our best to knit everything together well. Thanks a bunch for your contributions!

## and also

most probably addresses all issues people are having with

### ESP8266 arch pragma / yield definition woes
- [x] #114 and #119

### ESP8266 constructor initialization freezes
- [x] #29, #40, #53, #89, #113, #122

### ESP8266 watchdog timer (WDT)
- [x] #73, #81, #86, #101, #120

bogde commented:
I just tested HX711_full_example on Arduino Uno, WeMos D1 mini (ESP8266) and ESP32 DEVKIT V1 (ESP32). Everything seems to work as expected. I think it would be good to have some directions for testing though, what to test, how to test it, what to expect or look out for. We should probably have a dedicated thread for testing too. @amotl thank you for the time and effort you're putting into this. Thank you all!

bogde commented:
It also works on STM32F103C8T6 STM32 Blue Pill Board.

amotl commented:
Hi @bogde,

> I just tested HX711_full_example on [...]. Everything seems to work as expected. It also works on [...]. 

Good to hear, thank you so much for testing this on real iron. I will update the documentation reflecting your findings.

> I think it would be good to have some directions for testing though, what to test, how to test it, what to expect or look out for. We should probably have a dedicated thread for testing too.

Good idea. While I don't know if I'm the right guy for writing that, I would totally support if such a piece could be added to the documentation somehow. Having a dedicated thread about that topic would also make sense.

> @amotl thank you for the time and effort you're putting into this. Thank you all!

Appreciated. This would not have been possible without the various suggestions and pull requests from the community already, so really thanks to all!. We really just tried to put everything together again, we are happy that you like it.

Cheers,
Andreas.

amotl commented:
## Update

> Thank you so much for testing this on real iron. I will update the documentation reflecting your findings.

I've just updated the [README](https://github.com/hiveeyes/HX711/blob/spring-cleaning/README.md) to accordingly. Thanks again.

Also, we've added some things coming from #125 to work on giving users more choice on whether to wait for the HX711 chip or not when initializing the hardware.

rohlan commented on Makefile:
```diff
@@ -57,3 +57,6 @@ ci-all: install-platformio
 	$(platformio) ci --board=$(board) --lib="." examples/HX711_basic_example --verbose
+
+clean:
+	platformio run -t clean
```
$(...)? ;)

amotl replied:
Thanks for spotting this, @rohlan! I just pushed an update to the same branch, but @bogde merged this PR already. So, it will probably become part of eventual future updates.

amotl replied:
Made #126 from it. Thanks again.

amotl commented:
Thanks for merging, @bogde. It was a pleasure.

ClemensGruber commented:
I tested the lib succesful with an:
- Adafruit HUZZAH ESP8266
- Heltec WiFi Kit 32 (ESP32) 
- Adafruit Feather HUZZAH32 (ESP32)

amotl commented:
Thanks for testing on more hardware devices, @ClemensGruber! We just made #127 to account for that.

amotl commented:
@thiasB also [found](https://community.hiveeyes.org/t/improving-the-canonical-arduino-hx711-library-for-esp32-and-beyond/539/28) that
> Code is running fine on an Adafruit Feather M0 RFM95 Lora 1 (ATSAMD21G18 Cortex M0) board.

Thanks for testing!

amotl commented:
While @riemers also confirms more stability with the changes coming from here, he still reports about random erratic readings when running on ESP8266.

While he outlined his observations at https://github.com/Skaronator/ESP8266-Load-Cell/issues/8#issuecomment-469402617, we might might want to open a dedicated issue for that here.
